netwaystev2/src/filter/client_update.rs

```diff
@@ -1,29 +1,136 @@
 use anyhow::{anyhow, Result};
 use std::collections::{hash_map::Entry, HashMap};
 
+use super::{FilterNotice, FilterNotifySend};
+use crate::common::Endpoint;
 use crate::protocol::{GameUpdate, GenPartInfo, GenStateDiffPart, UniUpdate};
 use conway::{
     rle::Pattern,
-    universe::{GenStateDiff, Universe},
+    universe::{BigBang, GenStateDiff, PlayerBuilder, Region, Universe},
 };
 
 pub struct ClientRoom {
     player_name:       String, // Duplicate of player_name from OtherEndServer (parent) struct
+    player_id:         Option<usize>, // If player is not a lurker, must be Some(...) before the first GenStateDiff
+    other_players:     HashMap<String, Option<usize>>, // Other players: player_name => player_id (None means lurker)
     pub game:          Option<ClientGame>,
     pub last_chat_seq: Option<u64>, // sequence number of latest chat msg. received from server
 }
 
 pub struct ClientGame {
-    player_id:         usize,
+    player_id:         Option<usize>, // Duplicate of player_id from ClientRoom
     diff_parts:        HashMap<(u32, u32), Vec<Option<String>>>,
     universe:          Universe,
     pub last_full_gen: Option<u64>, // generation number client is currently at
     pub partial_gen:   Option<GenPartInfo>,
 }
 
 impl ClientRoom {
-    pub fn process_game_update(&mut self, game_update: &GameUpdate) -> Result<()> {
-        Ok(()) //XXX
+    pub async fn process_game_update(
+        &mut self,
+        server_endpoint: Endpoint,
+        game_update: &GameUpdate,
+        filter_notice_tx: &mut FilterNotifySend,
+    ) -> Result<()> {
+        use GameUpdate::*;
+        // First, special handling for some of these
+        match game_update {
+            GameStart { options } => {
+                let mut big_bang = BigBang::new()
+                    .width(options.width as usize)
+                    .height(options.height as usize)
+                    .server_mode(false)
+                    .history(options.history as usize)
+                    .fog_radius(options.fog_radius as usize);
+
+                for net_region in options.player_writable.iter() {
+                    let region: Region = net_region.clone().into();
+                    big_bang = big_bang.add_player(PlayerBuilder::new(region));
+                }
+                let uni = big_bang.birth()?;
+                let game = ClientGame {
+                    player_id:     self.player_id, // copy
```

Comment:
What's the comment for?

Reply:
It's copying the `player_id` from the room to the game.